### ISSUE TITLE
torch.compile max-autotune and reduce-overhead much slower than default due to repeated CUDA graph instantiation

### ISSUE THREAD

**Issue by trywithpython** [pytorch/pytorch#171672]
Suggested labels: module: inductor, module: cudagraph, perf  

### 🐛 Describe the bug

**Description:**

When compiling my model with `torch.compile`:

- `default` mode gives very good speed (~27 data/s).
- `max-autotune` and `reduce-overhead` modes are **much slower** (2~2.3 data/s).
- `max-autotune-no-cudagraphs` mode restores the speed back to normal (~27 data/s).

My model is mostly composed of Conv layers and does not have any special operations.

---

**Profiler observations:**

Using `torch.profiler`, I noticed that **in every iteration** the following CUDA Graph-related calls are executed:

```
cudaStreamEndCapture
cudaGraphInstantiateWithFlags
cudaGraphGetNodes
cudaGraphDestroy
cudaGraphLaunch
```

It looks like the CUDA graph is being **recreated every iteration** instead of reusing a captured graph.

Moreover, after `cudaGraphLaunch`, the execution switches to different streams each iteration,  
e.g., iteration 10 uses streams 590~599,  iteration 11 uses streams 600-609, etc.

---

**Expected behavior:**
CUDA graph capture should happen once and be reused across iterations.  
Having to reinstantiate and destroy the graph every iteration seems to cause major slowdown.

## 💬 Additional Notes

- `default` mode and `max-autotune-no-cudagraphs` do **not** suffer this slowdown.
- `max-autotune` and `reduce-overhead` **rebuild the CUDA Graph every iteration**.

Sincerely appreciate for your attention

### Versions

PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

cc @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng @chauhang

**Comment by Driller-Killer** [comment]
It seems that CUDA Graphs cannot be reused across different iterations. Is this due to the graph hitting a recapture condition?

**Comment by trywithpython** [comment]
> It seems that CUDA Graphs cannot be reused across different iterations. Is this due to the graph hitting a recapture condition?

Thanks, that matches our suspicion as well.
It looks like the CUDA Graph is being recaptured every iteration.

To clarify, in our setup:
- Input shapes are fixed across iterations.
- We do not explicitly switch or manage CUDA streams.
- There are no in-place ops or parameter/buffer mutations inside forward.

The only state change is that model parameters are updated between iterations as part of standard training (i.e., outside forward). Given that this is a standard training loop with fixed shapes, we would expect the captured CUDA Graph to be reusable across iterations.

Given this, could you help clarify what conditions would still force a CUDA Graph recapture (e.g., implicit stream changes, optimizer interactions, or allocator behavior)?
